plugins/outputs/iotdb/iotdb.go

```diff
@@ -229,6 +231,60 @@ func (s *IoTDB) convertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*rec
 	return rwt, nil
 }
 
+// takes in an array of compiled regex patterns
+// and validates them on the given tag.
+func matchToEncloseInBacktick(regexStrings []*regexp.Regexp, tag string) string {
+	// loops through all the regex patterns and if one
+	// pattern matches returns the tag between `
+	for _, regex := range regexStrings {
+		if regex.MatchString(tag) {
+			return "`" + tag + "`"
+		}
+	}
+
+	return tag
+}
+
+// checks is the tag contains any IoTDB invalid character
+func (s *IoTDB) validateTag(tag string) (string, error) {
+	// matches any word that has a non valid backtick
+	// `word`  							 <- dosen't match
+	// ``word , `wo`rd` , `word , word`  <- match
+	forbiddenBacktick, _ := regexp.Compile("^[^\x60].*?[\x60]+.*?[^\x60]$|^[\x60].*[\x60]+.*[\x60]$|^[\x60]+.*[^\x60]$|^[^\x60].*[\x60]+$")
+	allowedBacktick, _ := regexp.Compile("^[\x60].*[\x60]$")
+
+	// IoTDB uses "root" as a keyword and can be called only at the start of the path
+	if tag == "root" {
+		return "", errors.New("cannot use \"root\" as tag")
+	} else if forbiddenBacktick.MatchString(tag) { // returns an error if the backsticks are used in an inappropriate way
+		return "", errors.New("cannot use ` in tag names")
+	} else if allowedBacktick.MatchString(tag) { // if the tag in already enclosed in tags returns the tag
+		return tag, nil
+	}
+
+	regexArray := []*regexp.Regexp{} // array of compiled regex patterns
+
+	switch s.SanitizeTags {
+	case "0.13":
+		matchUnsopportedCharacter, _ := regexp.Compile("[^0-9a-zA-Z_:@#${}\x60]")
+
+		regex := []*regexp.Regexp{matchUnsopportedCharacter}
+		regexArray = append(regexArray, regex...)
+
+	// from version 1.x.x IoTDB changed the allowed keys in nodes
+	case "1.0", "1.1", "1.2", "1.3":
+		matchUnsopportedCharacter, _ := regexp.Compile("[^0-9a-zA-Z_\x60]")
+		matchNumericString, _ := regexp.Compile(`^\d+$`)
```

Comment:
Same here, use `MustCompile` and initialize `regexArray` in `Init()`!

Reply:
why initializing `regexArray` in `Init()`?
u think it would be better to include it in the struct for future changes?

and i would also delete those useless variables and directly append the regex to the array.
```golang
regexArray := []*regexp.Regexp{} // array of compiled regex patterns

	switch s.SanitizeTags {
	case "0.13":
		regex := []*regexp.Regexp{
			regexp.MustCompile("[^0-9a-zA-Z_:@#${}\x60]"),
		}
		regexArray = append(regexArray, regex...)

	// from version 1.x.x IoTDB changed the allowed keys in nodes
	case "1.0", "1.1", "1.2", "1.3":
		regex := []*regexp.Regexp{
			regexp.MustCompile("[^0-9a-zA-Z_\x60]"),
			regexp.MustCompile(`^\d+$`),
		}

		regexArray = append(regexArray, regex...)
	default:
		return tag, nil
	}
```
the idea is to make it easy to undertand and modify